### ISSUE TITLE
`think_format_reward` always returns 0.0 for models whose chat template prefills `<think>`

### ISSUE THREAD

**Issue by A-ryanVAT-S** [huggingface/trl#6966]
## Reproduction

On models whose chat template prefills `<think>` into the generation prompt, `think_format_reward`
returns 0.0 for every completion, including well-formed ones. No error, no warning.

The reward requires the completion to *start* with `<think>` (`trl/rewards/format_rewards.py`):

```python
pattern = r"^<think>(?!.*<think>)(.*?)</think>.*$"
```

But GRPO decodes completions from the generated tokens alone (`grpo_trainer.py:2761`), so when the
template already emitted `<think>` in the prompt, that tag never reaches the reward:

```python
from transformers import AutoTokenizer

from trl.rewards import think_format_reward

tok = AutoTokenizer.from_pretrained("deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B")
prompt = tok.apply_chat_template(
    [{"role": "user", "content": "What is 2+2?"}], tokenize=False, add_generation_prompt=True
)
print(repr(prompt[-24:]))  # prompt already ends with the opening tag

completion = "First, 2 plus 2.\n</think>\nThe answer is 4."  # so this is all the model generates
print(think_format_reward([[{"role": "assistant", "content": completion}]]))
```

outputs:

```
'+2?<｜Assistant｜><think>\n'
[0.0]
```

No download needed to check the template, since TRL bundles it.
`trl/chat_templates/deepseek_r1_distill.jinja` ends with:

```jinja
{% if add_generation_prompt and not ns.is_tool %}{{'<｜Assistant｜><think>\n'}}{% endif %}
```

Once the prompt ends in `<think>`, no completion can score 1.0: omit the tag and `^<think>` fails,
re-emit it and the `(?!.*<think>)` lookahead fails. In GRPO a constant reward gives zero advantage, so
the function contributes nothing for the whole run.

<details>
<summary>Confirmed with real generation (RTX 4060, greedy)</summary>

Completions sliced and decoded exactly as GRPO does, so the reward scores what the trainer would pass it:

```python
import torch
from transformers import AutoModelForCausalLM, AutoTokenizer

from trl.rewards import think_format_reward

PROMPT = [{"role": "user", "content": "What is 2+2? Think step by step, then give the answer."}]

for model_id in ["deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B", "Qwen/Qwen3-0.6B"]:
    tok = AutoTokenizer.from_pretrained(model_id)
    model = AutoModelForCausalLM.from_pretrained(model_id, dtype=torch.bfloat16).to("cuda")

    text = tok.apply_chat_template(PROMPT, tokenize=False, add_generation_prompt=True)
    inputs = tok(text, return_tensors="pt").to("cuda")
    prompt_len = inputs["input_ids"].shape[1]

    with torch.no_grad():
        out = model.generate(**inputs, max_new_tokens=512, do_sample=False)

    completion = tok.decode(out[0][prompt_len:], skip_special_tokens=True)  # what GRPO keeps

    print(f"=== {model_id} ===")
    print(f"  prompt tail        : {text[-40:]!r}")
    print(f"  completion (head)  : {completion[:110]!r}")
    print(f"  starts with <think>: {completion.lstrip().startswith('<think>')}")
    print(f"  contains </think>  : {'</think>' in completion}")
    print(f"  hit token limit    : {out[0][prompt_len:].shape[0] >= 512}")
    print(f"  think_format_reward: {think_format_reward([[{'role': 'assistant', 'content': completion}]])}")
```

outputs:

```
=== deepseek-ai/DeepSeek-R1-Distill-Qwen-1.5B ===
  prompt tail        : 'en give the answer.<｜Assistant｜><think>\n'
  completion (head)  : 'First, I recognize that the problem is asking for the sum of 2 and 2.\n\nNext, I perform the addition operation '
  starts with <think>: False
  contains </think>  : True
  hit token limit    : False
  think_format_reward: [0.0]

=== Qwen/Qwen3-0.6B ===
  prompt tail        : 'answer.<|im_end|>\n<|im_start|>assistant\n'
  completion (head)  : '<think>\nOkay, so the question is 2 plus 2. Let me think about how to approach this. First, I remember that whe'
  starts with <think>: True
  contains </think>  : True
  hit token limit    : False
  think_format_reward: [1.0]
```

Same prompt, same reward, same extraction. Only the template differs. The DeepSeek completion closes
with `</think>` and didn't hit the token limit, so this isn't truncation.

</details>

`qwen3_5_think.jinja:152` and `qwen3_6.jinja:152` carry the same branch, so I expect Qwen3.5/3.6 with
thinking enabled are affected too, though I read those templates rather than running the models.

For contrast, `reasoning_accuracy_reward` splits on `</think>` only and works fine here, which is why
the opening-tag anchor looks unintended rather than deliberate.

Flagging rather than patching, since the fix looks like a design call: relax the anchor, add a
parameter, or just document the constraint. Happy for anyone to pick it up.

## System Info

```
- Platform: Windows-11-10.0.26200-SP0
- Python version: 3.12.10
- TRL version: 1.13.0.dev0+6630e17
- PyTorch version: 2.13.0+cu130
- accelerator(s): NVIDIA GeForce RTX 4060 Laptop GPU
- Transformers version: 5.16.1
- Accelerate version: 1.14.0
- Accelerate config: not found
- Datasets version: 5.0.1
- HF Hub version: 1.29.0
- bitsandbytes version: not installed
- DeepSpeed version: not installed
- Liger-Kernel version: not installed
- PEFT version: not installed
- vLLM version: not installed
```

### Checklist

- [x] I have checked that my issue isn't already filed (see [open issues](https://github.com/huggingface/trl/issues?q=is%3Aissue))
- [x] I have included my system information
- [x] Any code provided is minimal, complete, and reproducible ([more on MREs](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any code provided is properly formatted in code blocks, (no screenshot, [more on code blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/creating-and-highlighting-code-blocks))
- [x] Any traceback provided is complete

**Comment by Ruinique** [comment]
I'll take this.

Plan: stop requiring the completion to start with `<think>` when the chat template already prefills that tag into the prompt, so GRPO-decoded completions (generated tokens only) still get a well-formed-format reward. I'll add a regression test for DeepSeek-R1-style prefilled templates.
